result/result.go

```diff
@@ -97,3 +100,16 @@ func (r Result[T]) UnwrapErr() error {
 
 	return r.err
 }
+
+// UnmarshalJSON implements the [json.Unmarshaler] interface.
```

Comment:
Do be consistent with option, please move the code and tests in `json.go` and `json_test.go` within the result package.